pageserver/src/walredo.rs

```diff
@@ -296,22 +295,22 @@ impl PostgresRedoManager {
                     n_attempts,
                     e,
                 );
-                // self.stdin only holds stdin & stderr as_raw_fd().
-                // Dropping it as part of take() doesn't close them.
-                // The owning objects (ChildStdout and ChildStderr) are stored in
-                // self.stdout and self.stderr, respsectively.
-                // We intentionally keep them open here to avoid a race between
-                // currently running `apply_wal_records()` and a `launch()` call
-                // after we return here.
-                // The currently running `apply_wal_records()` must not read from
-                // the newly launched process.
-                // By keeping self.stdout and self.stderr open here, `launch()` will
-                // get other file descriptors for the new child's stdout and stderr,
-                // and hence the current `apply_wal_records()` calls will observe
-                //  `output.stdout.as_raw_fd() != stdout_fd` .
-                if let Some(proc) = self.stdin.lock().unwrap().take() {
-                    proc.child.kill_and_wait();
+                // Avoid concurrent callers hitting the same issue.
+                // We can't prevent it from happening because we want to enable parallelism.
+                let mut guard = self.redo_process.lock().unwrap();
+                match &*guard {
+                    None => unreachable!("we never transition the field back to None"),
```

Comment:
Umm, you transition it back to None 4 lines below

Reply:
https://github.com/neondatabase/neon/pull/5557/commits/f59f97ac52fb751d6ff8ee08ced4a331b353a552